docs/csharp/whats-new/csharp-8.md

```diff
@@ -0,0 +1,314 @@
+
+```csharp
+int M()
+{
+    int y;
+    LocalFunction();
+    return y;
+
+    void LocalFunction() => y = 0;
+}
+```
+
+## Disposable ref structs
+
+A `struct` declared with the `ref` modifier may not implement any interfaces. Therefore, to enable a `ref struct` to be disposed, it need not implement <xref:System.IDisposable>, but it must have an accessible `void Dispose()` method.
+
+## Nullable reference types
+
+Inside a nullable annotation context, any variable of a reference type is considered to be a **nonnullable reference type**. If you want to indicate that a variable may be null, you must append the type name with the `?` to declare the variable as a **nullable reference type**.
+
+For nonnullable reference types, the compiler uses flow analysis to ensure that local variables are initialized to a non-null value when declared. Fields must be initialized to a non-null value using either an initializer or all constructors. Furthermore, nonnullable reference types can't be assigned a value that could be null.
+
+Nullable reference types aren't checked to ensure they aren't assigned or initialized to null. However, the compiler uses flow analysis to ensure that any variable of a nullable reference type is checked against null before it's accessed, or assigned to a nonnullable reference type.
+
+You can learn more about the feature in the overview of [nullable reference types](../nullable-references.md). Try it yourself in a new application in this [nullable reference types tutorial](../tutorials/nullable-reference-types.md). Learn about the steps to migrate an existing codebase to make use of nullable reference types in this tutorial for [migrating an application to use nullable reference types]../tutorials/upgrade-to-nullable-references.md).
+
+## Asynchronous streams
+
+Starting with C# 8.0, you can create and consume streams asynchronously. A method that returns an asynchronous stream has three properties:
+
+1. It was declared with the `async` modifier.
+1. It returns an `IAsyncEnumerable<T>`.
+1. The method contains `yield return` statements to return successive elements in the asynchronous stream.
+
+Consuming an asynchronous stream requires you to add the `await` keyword before the `foreach` keyword when you enumerate the elements of the stream. Adding the `await` keyword requires the method that enumerates the asynchronous stream was declared with the `async` modifier and returns a type allowed for an `async` method. Typically that means returning a <xref:System.Threading.Tasks.Task> or <xref:System.Threading.Tasks.Task%601>. It can also be a <xref:System.Threading.Tasks.ValueTask> or <xref:System.Threading.Tasks.ValueTask%601>. A method can both consume and produce an asynchronous stream, which means it would return an `IAsyncEnumerable<T>`.
+
+You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).
+
+## Ranges and indices
```

Comment:
This is a really useful feature, but I just don't get the syntax. It would be helpful to briefly describe the syntax of ranges and indices rather than just provide a set of examples. (Unless I'm just being obtuse...)

Reply:
I rewrote this entire section.  I do think I made it much harder to understand than it is.